### PR TITLE
feat: register and login form with validation

### PR DESCRIPTION
**Pull request by uzair-imtiaz**
### Description:
This pull request introduces frontend changes for the register and login forms. The Register component has been implemented using the form hook for form validation, enhancing user experience and improving data integrity. These changes aim to improve the user interface and provide a smoother registration and login process. The validation is in-sync with the real time input changes.

**Clone the Repository:**
`git clone git@github.com:Uzair-Imtiaz/code-book-frontend.git`
`cd code-book-frontend`

**Install Dependencies:**
`npm install`

**Checkout to the authentication branch:**
`git checkout authentication`

**Run the Frontend:**
`npm start`

**Register Form Testing:**
Open your web browser and navigate to http://localhost:3000/register.
Fill out the registration form with valid and invalid inputs.
Verify that validation errors are displayed for invalid inputs (e.g., empty fields, invalid email format, strong password requirement).
Verify that successful registration logs the data to the console.

### Snapshots:
Screenshot 2023-08-14 at 2 58 15 PM

Screenshot 2023-08-14 at 3 00 07 PM

### CODE REVIEW

File: src/components/authentication/Login.jsx [hunk 1]
```diff
@@ -0,0 +1,88 @@
+import React, { useState } from "react";
+import { Link } from 'react-router-dom';
+import "bootstrap/dist/css/bootstrap.css";
+import Logo from "../../images/logo.png";
+
+const Login = () => {
+  // State for storing username and password
+  const [username, setUsername] = useState('')
+  const [password, setPassword] = useState('')
```
Comment (Ali-D-Akbar) [inline src/components/authentication/Login.jsx]: Semicolons missing

File: src/components/authentication/Register.jsx [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        </div>
+      </div>
+    </div>
+
+  );
+};
+
+export default Register;
+
+// export default Register;
```
Comment (Ali-D-Akbar) [inline src/components/authentication/Register.jsx]: remove comment

File: src/index.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+import Register from './components/authentication/Register';
+import { createBrowserRouter, RouterProvider } from 'react-router-dom';
+
+const router = createBrowserRouter([
+  {
+    path: '/',
+    element: <Root />,
+    children: [
+
+    ],
```
Comment (Ali-D-Akbar) [inline src/index.js]: Can be kept in a single line

File: src/components/authentication/Login.jsx [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+  // Handle the form submission
+  const handleSubmit =(event) => {
+    console.log({username, password});
```
Comment (Ali-D-Akbar) [inline src/components/authentication/Login.jsx]: Remove this log statement. Exposes password.

File: src/components/authentication/Login.jsx [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+                </div>
+                <button type="submit" className="btn btn-primary ms-auto">
+                  Login
+                </button>
+              </form>
+            </div>
+          </div>
+          <div className="card-footer py-3 border-0">
+            <div className="text-center">
+              Don't have an account? <Link to="/register" className="text-dark">Create One</Link>
```
Comment (Ali-D-Akbar) [inline src/components/authentication/Login.jsx]: Create One -> Register (or Register Now)

File: src/components/authentication/Register.jsx [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+        errorMessage = 'Password must contain at least one lowercase letter. ';
+      }
+      if (!uppercaseRegex.test(value)) {
+        errorMessage = 'Password must contain at least one uppercase letter. ';
+      }
+      if (!digitRegex.test(value)) {
+        errorMessage = 'Password must contain at least one digit. ';
+      }
+      if (!specialCharRegex.test(value)) {
+        errorMessage = 'Password must contain at least one special character. ';
```
Comment (Ali-D-Akbar) [inline src/components/authentication/Register.jsx]: Remove extra spaces in the end of each message.

File: src/components/authentication/Register.jsx [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+  const isStrongPassword = (value) => {
+    const lowercaseRegex = /[a-z]/;
+    const uppercaseRegex = /[A-Z]/;
+    const digitRegex = /\d/;
+    const specialCharRegex = /[!@#$%^&*()_+{}\[\]:;<>,.?~\-\\/]/;
+
+    let errorMessage = '';
+
+    if (value.length < 8) {
+      errorMessage = 'Password must be at least 8 characters long';
```
Comment (Ali-D-Akbar) [inline src/components/authentication/Register.jsx]: either add a fullstop at every message or don't put at all.

File: src/components/authentication/Register.jsx [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+        errorMessage = 'Password must contain at least one uppercase letter. ';
+      }
+      if (!digitRegex.test(value)) {
+        errorMessage = 'Password must contain at least one digit. ';
+      }
+      if (!specialCharRegex.test(value)) {
+        errorMessage = 'Password must contain at least one special character. ';
+      }
+    }
+    return errorMessage === '' ? true : errorMessage;
```
Comment (Ali-D-Akbar) [inline src/components/authentication/Register.jsx]: Why not return multiple errors?

Comment (Ali-D-Akbar) [reply src/components/authentication/Register.jsx]: Also, return something like a json object to use like this:

```
return {
    valid: errors.length === 0,
    errors: errors,
  };
```

File: src/components/authentication/Register.jsx [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+const Register = () => {
+  const {
+    register,
+    handleSubmit,
+    formState: { errors },
+    watch,
+  } = useForm({ mode: 'onChange' });
+
+  const onSubmit = (data, event) => {
+    console.log(data);
```
Comment (Ali-D-Akbar) [inline src/components/authentication/Register.jsx]: remove console log

**Review by Ali-D-Akbar — CHANGES_REQUESTED** [review]
Can you also configure a standard eslint config in your code so your code is following eslint as expectedly?

File: src/components/authentication/Register.jsx (hunk 8 above)
Comment (uzair-imtiaz) [reply src/components/authentication/Register.jsx]: > Why not return multiple errors?

The error message gets too long if they are appended and they are to be displayed under the input field.

Comment (uzair-imtiaz) [reply src/components/authentication/Register.jsx]: > Also, return something like a json object to use like this:
> 
> ```
> return {
>     valid: errors.length === 0,
>     errors: errors,
>   };
> ```

is it usable with the validate property inside the register function?

Comment (Ali-D-Akbar) [reply src/components/authentication/Register.jsx]: I'd rather have a list of requirements for passwords before I'm done selecting it. It's discouraging to see that I haven't used a capital letter and then seeing that I don't have a number. Later seeing that I miss an underscore. You know what I mean?

> is it usable with the validate property inside the register function?

You gotta check that. I'm suggesting this return format so we're not expecting different return types from a function (like we do right now, it's either bool or string)

Comment (uzair-imtiaz) [reply src/components/authentication/Register.jsx]: It seems that the validation function has to return true or a string. The returned string is displayed then. 
About the error messages I'll change that.

File: src/components/authentication/Login.jsx [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+      setState({ value, error: true });
+    }
+    else {
+      setState({ value, error: false });
+    }
+  };
+
+  // Handle the form submission
+  const handleSubmit = (event) => {
+    console.log({ username, password });
```
Comment (Ali-D-Akbar) [inline src/components/authentication/Login.jsx:28]: please remove all the console log statements. I asked you to remove these before too.

**Review by Ali-D-Akbar — APPROVED** [review]

**Pull request merged**